processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -152,66 +152,87 @@ func (tsp *tailSamplingSpanProcessor) samplingPolicyOnTick() {
 		}
 		trace := d.(*sampling.TraceData)
 		trace.DecisionTime = time.Now()
-		for i, policy := range tsp.policies {
-			policyEvaluateStartTime := time.Now()
-			decision, err := policy.Evaluator.Evaluate(id, trace)
-			stats.Record(
-				policy.ctx,
-				statDecisionLatencyMicroSec.M(int64(time.Since(policyEvaluateStartTime)/time.Microsecond)))
-			if err != nil {
-				trace.Decisions[i] = sampling.NotSampled
-				evaluateErrorCount++
-				tsp.logger.Error("Sampling policy error", zap.Error(err))
-				continue
+
+		decision, policy, sampleCount, notSampleCount, errorCount := tsp.makeDecision(id, trace)
+		decisionSampled += sampleCount
+		decisionNotSampled += notSampleCount
+		evaluateErrorCount += errorCount
+
+		// Sampled or not, remove the batches
+		trace.Lock()
+		traceBatches := trace.ReceivedBatches
+		trace.ReceivedBatches = nil
+		trace.Unlock()
```

Comment:
Previously the lock was taken twice when a policy matched, I just combined them into a single acquisition